src/analysis/refactor_extract_region.ml

```diff
@@ -0,0 +1,395 @@
+open Std
+
+exception Nothing_to_do
+exception Not_allowed_in_interface_file
+
+let () =
+  Location.register_error_of_exn (function
+    | Nothing_to_do -> Some (Location.error "Nothing to do")
+    | Not_allowed_in_interface_file ->
+      Some
+        (Location.error
+           "Expression extraction is only allowed in implementation file")
+    | _ -> None)
+
+module Fresh_name = struct
+  (* Generate a fresh name that does not already exist in given environment. *)
+  let gen_val_name basename env =
+    let rec loop n =
+      let guess = basename ^ Int.to_string n in
+      if Env.bound_value guess env then succ n |> loop else guess
+    in
+    loop 1
+end
+
+module Gen = struct
+  let unit = Longident.Lident "()" |> Location.mknoloc
+
+  (* Generates [let name = body]. *)
+  let toplevel_let ~name ~body =
+    let open Ast_helper in
+    let pattern = Pat.mk (Ppat_var { txt = name; loc = Location.none }) in
+    let body = Parsetree_utils.filter_expr_attr body in
+    Str.value Nonrecursive [ Vb.mk pattern body ]
```

Comment:
Looking to `Nonrecursive`.
I am a little but curious about case like that: 

```ocaml
let f = 
  let x = 1 in 
  let y = 2 in 
  let z = x + y in 
  z + 1
```

Let's imagine that we extract `let z = x + y in` or `let z = x + y` we can imagine that it will produce the following code: 

```ocaml
let z x y = x + y

let f = 
  let x = 1 in 
  let y = 2 in 
  (z x y) + 1
```

No? In other terms, I think that if the selected range is a let-binding, we change the binding as a toplevel value (and we preserve the name if it makes senses, ie: if the binding is free, using Fresh_name, or using a fresh binding) 

So we can imagine that if we extract a recursive definition, we preserve it.